lib/spack/spack/detection/path.py

```diff
@@ -72,13 +74,146 @@ def executables_in_path(path_hints=None):
+    search_paths = llnl.util.filesystem.search_paths_for_libraries(*path_hints)
+
+    path_to_lib = {}
+    # Reverse order of search directories so that a lib in the first
+    # LD_LIBRARY_PATH entry overrides later entries
+    for search_path in reversed(search_paths):
+        for lib in os.listdir(search_path):
+            lib_path = os.path.join(search_path, lib)
+            if is_readable(lib_path):
+                path_to_lib[lib_path] = lib
+    return path_to_lib
+
+
 def _group_by_prefix(paths):
     groups = collections.defaultdict(set)
     for p in paths:
         groups[os.path.dirname(p)].add(p)
     return groups.items()
 
 
+# TODO consolidate this with by_executable
+# Packages should be able to define both .libraries and .executables in the future
+# determine_spec_details should get all relevant libraries and executables in one call
+def by_library(packages_to_check, path_hints=None):
+    # Techniques for finding libraries is determined on a per recipe basis in
+    # the determine_version class method. Some packages will extract the
+    # version number from a shared libraries filename.
+    # Other libraries could use the strings function to extract it as described
+    # in https://unix.stackexchange.com/questions/58846/viewing-linux-library-executable-version-info
+    """Return the list of packages that have been detected on the system,
+    searching by LD_LIBRARY_PATH.
+
+    Args:
+        packages_to_check (list): list of packages to be detected
+        path_hints (list): list of paths to be searched. If None the list will be
+            constructed based on the LD_LIBRARY_PATH environment variable.
+    """
+    path_to_lib_name = libraries_in_ld_library_path(path_hints=path_hints)
+    lib_pattern_to_pkgs = collections.defaultdict(list)
+    for pkg in packages_to_check:
+        if hasattr(pkg, 'libraries'):
+            for lib in pkg.libraries:
+                lib_pattern_to_pkgs[lib].append(pkg)
+
+    pkg_to_found_libs = collections.defaultdict(set)
+    for lib_pattern, pkgs in lib_pattern_to_pkgs.items():
+        compiled_re = re.compile(lib_pattern)
+        for path, lib in path_to_lib_name.items():
+            if compiled_re.search(lib):
+                for pkg in pkgs:
+                    pkg_to_found_libs[pkg].add(path)
+
+    pkg_to_entries = collections.defaultdict(list)
+    resolved_specs = {}  # spec -> lib found for the spec
+
+    for pkg, libs in pkg_to_found_libs.items():
+        if not hasattr(pkg, 'determine_spec_details'):
+            llnl.util.tty.warn(
+                "{0} must define 'determine_spec_details' in order"
+                " for Spack to detect externally-provided instances"
+                " of the package.".format(pkg.name))
+            continue
+
+        for prefix, libs_in_prefix in sorted(_group_by_prefix(libs)):
+            # TODO: multiple instances of a package can live in the same
```

Comment:
You can remove this comment since it's redundant WRT the `by_executable` function.